### PR TITLE
Added brief on DOM interfaces + func signatures + support hash link to notes

### PR DESCRIPTION
**Pull request by lakbychance**
Change log:-
1) Added a Note on DOM interfaces on top of DOM Manipulation section.
2) Updated function signatures of methods discussed.
3) Updated querySelector and querySelectorAll links from type Element to Document.

@bendtherules

### CODE REVIEW

**Comment by bendtherules** [comment]
Good job on making things nore consistent. 👌

So, basically when we say something is defined on node, it should be available on all the 3 places - doc, ele and docfrag. Is that correct understanding?  

1. Should we also add a line in first note for this - after "inherit from node", saying like "So when we define node.abc, that means it's available on all 3"? Explicit is generally better, but if it is obvious, then let's skip it. 

2. Shouldn't querySelector and querySelectorAll also be defined on Node interface, then? (i guess this is available on docfrag also)

3. In docfrag section, i am slightly against moving appendChild note to comments - because that is very hard to read.  
  If we are talking about a concept like docfrag, we should also say there how to actually use it. Without the appendChild text, someone might get stuck on how do i finally add it back to dom? It's not super obvious that appendChild can append multiple child also.

**Comment by lakbychance** [comment]
Thanks 😄 

_1) So, basically when we say something is defined on node, it should be available on all the 3 places - doc, ele and docfrag. Is that correct understanding?_

**I think so. Because in the browser console, I added few properties to Node.prototype and was able to access the same on document, element and document fragment.**  

_2) Shouldn't querySelector and querySelectorAll also be defined on Node interface, then? (i guess this is available on docfrag also)_

**querySelector & querySelectorAll are actually available via ParentNode mixin. Then I searched what's a mixin. So got to know that instead of inheriting from Node (which doesn't have querySelector and querySelectorAll btw), Document, Element and DocumentFragment includes this ParentNode mixin. So I guess composition was chosen instead of inheritance for some of the features.** 

_3) In docfrag section, i am slightly against moving appendChild note to comments - because that is very hard to read.
If we are talking about a concept like docfrag, we should also say there how to actually use it. Without the appendChild text, someone might get stuck on how do i finally add it back to dom? It's not super obvious that appendChild can append multiple child also._

**I thought of moving this appendChild just due to consistency with all the other methods but I guess we can restate it at original place for clarity.**

File: docs/create-modify-element.mdx [hunk 1]
```diff
@@ -9,6 +9,12 @@ import Admonition from "../components/Admonition"
 import Accordion from "../components/Accordion"
 import MDN from "../components/MDNBadge"
 
+:::note DOM interfaces
+Every kind of **DOM** node is represented by an interface based on **Node** interface.   
+**Element, Document and DocumentFragment** interfaces inherit from Node.  
+More about DOM interfaces [here](https://developer.mozilla.org/en-US/docs/Web/API/Document_Object_Model)
```
Comment (bendtherules) [inline docs/create-modify-element.mdx]: Don't put everything in newline, can be distracting to readers. 
I understand that 2nd line is separate to draw reader's attention. Then, let's put 2nd and 3rd line together?

File: docs/create-modify-element.mdx [hunk 2]
```diff
@@ -229,6 +233,8 @@ outerEle.appendChild(fragment)
 // This will remove all content from fragment and 
 // insert them within outerEle.
 
+// .appendChild()/.insertBefore() - can be used to move the whole content of the fragment to the actual DOM.
```
Comment (bendtherules) [inline docs/create-modify-element.mdx]: Let's add this as note (within details) in appendChild section? Then, we can link back to that note from here.

**Comment by bendtherules** [comment]
1,2. It seems like difference between Node and ParentNode is almost a specification detail.  

Example - Node has `.appendChild`, which logically only makes sense on ParentNode interface - but it is not there. Also, on the `Node` page, it says -
> a particular feature of the base Node interface may not apply to one of its child interfaces; in that case, the inheriting node may return null or throw an exception

meaning some node methods are actually not available/useful on all type of nodes.  

All of these makes me think, given our goals - we **shouldn't distinguish between Node and ParentNode** - and if some method is available on these 3 major types of nodes, we can say it is available on `Node` directly. (We can explain this assumption in the first note you added.)  

Other option can be writing the exact interface on which it is defined, like `ParentNode.qs`, but linking to a specific implementation.

Which one do you think is better?

3. - https://github.com/bendtherules/UiQuestions/pull/3#discussion_r510707769

File: docs/create-modify-element.mdx (hunk 1 above)
Comment (lakbychance) [reply docs/create-modify-element.mdx]: Yup that can be done.

File: docs/create-modify-element.mdx (hunk 2 above)
Comment (lakbychance) [reply docs/create-modify-element.mdx]: Sorry didn't get this. Where exactly line 236 content should be put in appendChild ? And also what should be done for insertBefore ? How do you plan to link back to that note ?

**Comment by lakbychance** [comment]
@bendtherules 
_All of these makes me think, given our goals - we shouldn't distinguish between Node and ParentNode - and if some method is available on these 3 major types of nodes, we can say it is available on Node directly. (We can explain this assumption in the first note you added.)_

This can be done but **Node.prototype** doesn't have **querySelector** and **querySelectorAll**. **Element.prototype**, **Document.prototype** and **DocumentFragment.prototype** - all three have querySelector and querySelectorAll. 

In our docs, we are actually showcasing document's implementation and ideally for interview purposes that should be enough. So I am thinking whether to bring **ParentNode** into picture at all ? Maybe add a collapsable note (Related) in there with ParentNode specifics and link to MDN ParentNode page ?

**Comment by bendtherules** [comment]
Cool, let's just write document.qs in signature, but also mention that this is available on individual elements and documentFragment. Sounds good?

File: docs/create-modify-element.mdx (hunk 2 above)
Comment (bendtherules) [reply docs/create-modify-element.mdx]: Let me make this change, it'll be easier to show that way. (i meant #abc style link within same page, if we can somehow add a anchor within the note)

**Comment by lakbychance** [comment]
@bendtherules 
_Cool, let's just write document.qs in signature, but also mention that this is available on individual elements and documentFragment. Sounds good?_

Did this in the recent commit. 
Haven't updated the link back stuff that you are suggesting since can't picture it properly.

File: docs/create-modify-element.mdx (hunk 2 above)
Comment (lakbychance) [reply docs/create-modify-element.mdx]: Cool 👍

**Comment by bendtherules** [comment]
Thanks. I'll try to do that one.

**Comment by bendtherules** [comment]
@lapstjup 
I have added few commits in your branch (which is a nice permission in GH 👍 ) to link back to notes. Can you try out the [deploy preview](https://deploy-preview-3--uiquestions.netlify.app/) and see how it feels?

Changes -
1. From `qsAll`, link back to `qs` notes. From fragment, link back to `appendChild` notes. Links use the anchor tag / id mechanism.
2. \<details/> tags don't open automatically if you link to a id within it 😢 . So, added custom solution in Accordion component to auto-open itself if hash in url matches id of something within it. This is not as good as default browser behaviour, but kinda works. (this is what needs testing)

**Comment by lakbychance** [comment]
@bendtherules 
I tried the new changes. Few observations:-
- Clicking on the link like `use appendChild` first time, the appendChild note opens in a nice manner due to the `openAccordionOnHash()` logic where `scrollIntoView` is used.
- Now if the note is already opened and I click on `use appendChild` again, the `openAccordionOnHash()` doesn't trigger because there isn't any `hashChange` event. But since it's a id fragment, the default behaviour of going to that fragment happens. But in this case, some part of the top of note get's hidden under the navbar and is partially visible to user and also it's not in center as in `scrollIntoView`,
- Now if the same note was closed manually and i click on `use appendChild` again, the `openAccordionOnHash()` doesn't trigger because there isn't any `hashChange` event. But also default behaviour of going to that fragment doesn't happen well because as you stated `<details/>` doesn't open automatically if you have id within it.


To address the same, I have a proposal (not the best) but maybe it gives you a better idea:-

```
const Link = ({hash,title}) =>{

    const handleLink=(event)=>{
        event.preventDefault();
        window.location.hash  = "#";
        window.location.hash = hash;
      }
      
    return <a onClick={handleLink} className="navlink">{title}</a>
}
```
This `Link` component can be used instead of markdown links. And so hashChange will always trigger and also, scrollIntoView will always be the one used to bring the content into focus instead of default anchor tag behaviour.

**Comment by bendtherules** [comment]
Yes, i saw the same behaviors. One more thing is what happens when you refresh with hash in url - currently it expands, but atleast chrome ignores the scrollIntoView thing - preserving the last scroll pos (which is actually better /default behavior).

Like you said, then we have to go deeper and handle it in \<a> layer. But now we have more potential problems -

1. Surely, docusaurus uses its own Link component (connected with router) to handle markdown urls. So, we can't just use a tag.

2. What if i refresh/open the page with hash in url? On direct link open, it should still open and scroll to that element. So, we need to also handle it at url level.

**Comment by lakbychance** [comment]
@bendtherules 
What if we create a wrapper component and wrap content of `create-modify-element` inside it and do some of the stuff to tackle above issues as you stated ?

Note - I have prefixed the reference links like of `element-has-qs` with `link`. It can be something else too. This is to differentiate between docusauraus links and our reference links within doc.

```
const DocWrapper = ({children})=>{

   useEffect(()=>{
    let hash = location.hash.substring(1);
    // Remove scroll restoration when there is a hash of length > 0.
    if (hash.length!==0 && "scrollRestoration" in history) {
      history.scrollRestoration = "manual"
    }
    // Check if the hash type is a Docusauraus link or a reference link within document.
    let isDocusaurusLink = (hash.substring(0,4) !== "link")
    let targetElement;
    targetElement = document.getElementById(hash)
    if(!targetElement)return
    //if link is Docusaurus link - use scroll Option as true else if link is a reference to something within
    // then use a smooth scroll with center block
    const scrollOptions = !isDocusaurusLink?{behavior:"smooth",block:"center"}:true
    targetElement.scrollIntoView(scrollOptions)
    if(!isDocusaurusLink)
    targetElement.focus();
   },[])

   return <>{children}</>
}
```

**Comment by bendtherules** [comment]
If we wrap the whole mdx in a component, we'll lose all benefit of markdown. Like you saw, we can't use plain markdown syntax reliably inside any tag like details.

**Comment by lakbychance** [comment]
After the import of DocWrapper in create-modify-element.mdx we can wrap the markdown. It's working for me. Are you saying this might break something in future ?

**Comment by bendtherules** [comment]
Does this render correctly?

```
<DocWrapper>
*This* is `some code` in text. 
</DocWrapper>
```

**Comment by lakbychance** [comment]
This renders correctly:-

```
<DocWrapper>

*This* is `some code` in text. 

</DocWrapper>
```

Those newlines are needed.

**Comment by bendtherules** [comment]
Hmm interesting, i thought all markdown syntax doesn't work correctly within tags (specially that extra `:::note` syntax).

If all that is working, then great 👍. You can push to this branch or create new Pr - so I'll also be able to try it out.

**Comment by lakbychance** [comment]
@bendtherules 
You can check the changes

**Comment by bendtherules** [comment]
The solution is working. Very cool.

But I have a few doubts -
1. Which code is solving what?

    We had **2 types of problems** mainly earlier - A. click again doesn't trigger, B. doesn't scroll on page load.  
    To solve A, custom Link changes the hash twice - understandable.  
    To solve B,  added DocWrapper - which does 2 things - disables scrollRestoration (understandable) and **scrollIntoView (why?)**.  
    
    **Why is DocWrapper scrolling?** - Is it to scroll to hash during initial load? Because if that is the reason, `<Accordion>` is already doing that (it sets evt handler for hashChange + also calls it once initially). Further expanding on this, do we really need separate DocWrapper then? We can move the scrollRestoration changing code to Accordion itself.

2. Link component is not accessible. It should have the hash in href attribute (for ex. to open in new tab). Also, let's keep the prop names same as \<a/> tag and call it InternalLink or something.

3. I liked the idea of `link-` to disambiguate section links and this lookback links, but not sure if we actually need it. If some link is within a Accordion, we need to open/scroll to it - shouldn't matter what kind of link it is.

**Comment by lakbychance** [comment]
> Which code is solving what?
> 
> We had 2 types of problems mainly earlier - A. click again doesn't trigger, B. doesn't scroll on page load.
> To solve A, custom Link changes the hash twice - understandable.
> To solve B, added DocWrapper - which does 2 things - disables scrollRestoration (understandable) and scrollIntoView (why?).

scrollIntoView because - the page doesn't scroll by default to the section links on page reload if the section link hash was in the url. Just disabling scrollRestoration doesn't work. That's why I am manually picking up the targetElement and scrolling over there.

> Why is DocWrapper scrolling? - Is it to scroll to hash during initial load? Because if that is the reason, <Accordion> is already doing that (it sets evt handler for hashChange + also calls it once initially). Further expanding on this, do we really need separate DocWrapper then? We can move the scrollRestoration changing code to Accordion itself.

The Accordion logic is the same as you made and I think that it would be better that to let it not handle scrollRestoration, firstly because of the section links stuff and secondly, it's an initial load operation and not tied with Accordions.

> Link component is not accessible. It should have the hash in href attribute (for ex. to open in new tab). Also, let's keep the prop names same as <a/> tag and call it InternalLink or something.

Yeah makes sense. Instead of `title` prop in `Link`, should we use `children` in `InternalLink` to give same feel as a tag besides the href change?

> I liked the idea of link- to disambiguate section links and this lookback links, but not sure if we actually need it. If some link is within a Accordion, we need to open/scroll to it - shouldn't matter what kind of link it is.

If we don't want to direct to section links on page reload if the hash is in url, then we don't need that link prefix. Then maybe scrollRestoration can also be part of Accordion. It's really to handle these two types of links.

**Comment by bendtherules** [comment]
Ok got it. 

So the question is - do we want to change scroll Restoration behavior for section links or not? (if unchanged, it is the default browser behavior - but different from special link handling; vice versa)

IMO, we shouldn't change default browser behavior as much as possible. That means, on reload with section hash, scroll position will be same as before reload - it will not scroll to section. But on initial load with section hash, it should scroll to section.

For the special link also, we should try to have same behavior. But i dont think its easy to identify if it's a reload vs initial load. So, we can always assume initial load and open/scroll to them always.

(in short, let's not modify scroll behavior for section links, only special links)

**Comment by bendtherules** [comment]
I will try to make those changes today - to see if it works and if we can do with less code.

**Comment by bendtherules** [comment]
@lapstjup Ok, I get what you were saying now. When we set `.scrollRestoration`to 'manual', that means we will take care restoring scroll from **next** route changes - not current one. That means, we are not only handling it for urls with special link, but any url change in our app. 

That means, at the very least we need to make sure that section hashes scroll (which you did). But the problem is - this is NOT the only case we have to handle. We also have to then restore scroll if you are on page1, scroll a bit, go to page2 and then go back in history to page1 (which we are not doing currently).

Given that with this approach, we have to handle all kind of scroll restoration - I am thinking whether it is even worth the effort. We only want to handle scroll for special hashes. 

1. The prev approach i did for this was to explicitly scroll on page load, but that is ignored due to scroll restoration. Maybe we can try that with small setTimeout??

2. OR, maybe we change `.scrollRestoration` to "manual" conditionally on InternalLink click, handle it on load, then set it back to "auto"?

**Comment by lakbychance** [comment]
@bendtherules Happy Dussehra firstly 😄 

Now back to the stuff:- 

I think setTimeout one might work out. If we start with handling scroll restoration for the internal links, there will eventually be too much to handle with it. So if using setTimeout with a time like 100 or so milliseconds works for us, I think that would be good. 
In this case, we only scroll to our internal links after that timeout.

If we want to really handle scroll restoration in future (which seems like a rare scenario unless we have different types of hash links to handle), we would need to rethink some kind of reusable and robust logic for that which checks cross browser compatibility as well. 

Also in this case the distinguishing between section links and Internal Links won't matter I guess because, for section links we are not defining anything new. We are going to leave that to browser. The timeout can probably exist inside `openAccordionOnHash` after the `contains` is valid.

so we can do this like so:-

```
 function openAccordionOnHash() {
    const hash = location.hash.substring(1);
    if (hash.length === 0) {
      return;
    }
    let targetElement;
    targetElement = document.getElementById(hash);
    if (targetElement === null) {
      return;
    }

    // 1. If target is within current details element
    const detailsElement = detailsRef.current;
    if (
      detailsElement !== null &&
      detailsElement.contains(targetElement)
    ) {
      // 2. open details panel with a timeout to work with scroll restoration
      setTimeout(()=>{
        if (!detailsElement.open) {
          detailsElement.open = true;
        }
        // 3. and scroll into view and focus
        targetElement.focus();
        targetElement.scrollIntoView({ behavior: "smooth", block: "center" });
      },100)
  
    }
  }
```

We can remove link prefixes, DocWrapper also since they are redundant now.

**Comment by bendtherules** [comment]
@lapstjup Happy Dashami (It's actually one day later here 😀)

I read about the scrollRestoration thing a little more (from SO and the proposal), to understand exactly at which point they check the value to decide scroll.  
It's like - wait for whole page load (head, body render), \<some potential delay\>, **now check scrollRestoration value**, then restore last state. It's *NOT* read before page load.  

So, we have the opportunity to change value to "manual" while page is loading (checking hash). That means, we can also handle both first load + reloads - basically, conditionally change the restoration value during page load. After we scroll ourselves, we can set back value to default to let browser handle others.

So, I made these changes -
1. Remove DocWrapper
2. Add handleScrollRestoration script to head - if hash starts with "#link-", set to "manual".
3. Accordion - Scroll & then restore value to "auto"  
    (Needs 1sec delay for mobile browsers. If we do this too quickly, by the time browser checks - it is again back to "auto" - so it will scroll again after our scroll.)

4. Link renamed to InternalLink

I have made the changes on a [different branch](https://github.com/bendtherules/UiQuestions/tree/fix-details-scroll) - which is [deployed here](https://fix-details-scroll--uiquestions.netlify.app/). Can you please check this new version?

(There is one small issue - if you open a Accordion above, then click on a section below, then refresh - the auto scroll position is not same as before refresh, it moves up by some amount. I think this is because the Accordion is closed now and so, doc height/scroll pos has changed.)

**Comment by bendtherules** [comment]
On further thoughts, it striked me that there could have been a simpler solution - On hash link click or page open, we just open the \<details/> element - and that's all, browser should then scroll by itself.

But that has the irritating top cut off problem - because of the topbar. Looks like docusaurus somehow fixed this for section links (need to check how), but that same handling doesn't work for non-section ids (like ours). I used their own \<Link/> component just to make sure.

I basically just tried this -
```jsx
  const handleLink = () => {
    const hash = to.substring(1);
    const targetElement = document.getElementById(hash);
    if (targetElement !== null) {
      const parentDetails = targetElement.closest("details");
      if (parentDetails !== null) {
        parentDetails.open = true;
      }
    }
  };

  return <Link onClick={handleLink} to={to} {...otherProps}></Link>;
```

Unless that is fixed in docusaurus land, our current solution is IMO better (though more complex).

**Comment by lakbychance** [comment]
@bendtherules 
I tested your changes. Yes scroll restoration works well. That last bit of page not restoring to exact scroll position is only when the accordion is open before the page refresh gets triggered. 

Just one doubt though - Is there a downside to the setTimeout way of doing opening the Accordion ? Because we don't need the link prefixes and dealing with scrollRestoration in that bit ?

**Comment by bendtherules** [comment]
Didn't understand your last question completely.

There scrolling is instant, but resetting scroll Restoration is after 1 sec. 

1. Yes, that means there is a small chance that someone can go back in that time and scroll restoration wouldn't work. Did you mean that?

2. Another edge case - if something is a "#link-", but doesn't have a parent Accordion - then also scrollRestoration will never get back to "auto". Is this what you meant?

Now that I think about it - this is only solving for hash within Accordion - opening Accordion and scrolling to hash correctly are separate problems, but their solution order has to be correct. Open Accordion, then scroll.
(maybe we can take this up later)

**Comment by lakbychance** [comment]
@bendtherules I mean this bit. Is there a significant downside to this ?
> ```
>  function openAccordionOnHash() {
>     const hash = location.hash.substring(1);
>     if (hash.length === 0) {
>       return;
>     }
>     let targetElement;
>     targetElement = document.getElementById(hash);
>     if (targetElement === null) {
>       return;
>     }
> 
>     // 1. If target is within current details element
>     const detailsElement = detailsRef.current;
>     if (
>       detailsElement !== null &&
>       detailsElement.contains(targetElement)
>     ) {
>       // 2. open details panel with a timeout to work with scroll restoration
>       setTimeout(()=>{
>         if (!detailsElement.open) {
>           detailsElement.open = true;
>         }
>         // 3. and scroll into view and focus
>         targetElement.focus();
>         targetElement.scrollIntoView({ behavior: "smooth", block: "center" });
>       },100)
>   
>     }
>   }
> ```
> 
> We can remove link prefixes, DocWrapper also since they are redundant now.

**Comment by bendtherules** [comment]
I actually haven't merged the new code from that branch (for which you saw the preview). setTimeout is not there for opening there. Check the linked branch above.

Or do you mean we should do the quoted thing? I think we shouldn't - any delay means page will render, then scroll. Without delay, it works like native chrome - scroll, then render scrolled page.

**Comment by lakbychance** [comment]
@bendtherules 
Nevermind. That setTimeout approach was actually tried by me and though it works and doesn't need link prefixes and scrollRestoration api, it's not good to bet on a 100 seconds timeout or something. 

Let's go with your changes.

Just also include the InternalLink for `use appendChild`. I think that is missing.

Let me know when you do that and I will close this PR.

**Comment by bendtherules** [comment]
Cool, i will add for that one and finally merge this PR.

(this is the longest discussion i had on gh 😅)

**Pull request merged**